internal/services/mongocluster/mongo_cluster_firewall_rule_resource.go

```diff
@@ -0,0 +1,225 @@
+			}
+
+			mongoClusterId, err := firewallrules.ParseMongoClusterID(state.MongoClusterId)
+			if err != nil {
+				return err
+			}
+
+			id := firewallrules.NewFirewallRuleID(mongoClusterId.SubscriptionId, mongoClusterId.ResourceGroupName, mongoClusterId.MongoClusterName, state.Name)
+
+			existing, err := client.Get(ctx, id)
+			if err != nil {
+				if !response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+				}
+			}
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			properties := firewallrules.FirewallRule{
+				Properties: &firewallrules.FirewallRuleProperties{
+					EndIPAddress:   state.EndIpAddress,
+					StartIPAddress: state.StartIpAddress,
+				},
+			}
+
+			if err := client.CreateOrUpdateThenPoll(ctx, id, properties); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+
+			return nil
+		},
+	}
+}
+
+func (r MongoClusterFirewallRuleResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.MongoCluster.FirewallRulesClient
+
+			id, err := firewallrules.ParseFirewallRuleID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(*id)
+				}
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			state := MongoClusterFirewallRuleResourceModel{
+				Name:           id.FirewallRuleName,
+				MongoClusterId: firewallrules.NewMongoClusterID(id.SubscriptionId, id.ResourceGroupName, id.MongoClusterName).ID(),
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					state.EndIpAddress = props.EndIPAddress
+					state.StartIpAddress = props.StartIPAddress
+				}
+			}
+
+			return metadata.Encode(&state)
+		},
+	}
+}
+
+func (r MongoClusterFirewallRuleResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.MongoCluster.FirewallRulesClient
+
+			id, err := firewallrules.ParseFirewallRuleID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var state MongoClusterFirewallRuleResourceModel
+			if err := metadata.Decode(&state); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			if existing.Model == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", *id)
+			}
+
+			model := *existing.Model
+
+			if metadata.ResourceData.HasChange("end_ip_address") {
+				if model.Properties == nil {
```

Comment:
If `Properties` was nil, we should error since the fields in that struct are required. Then we don't also have to duplicate this check in both `HasChange` checks

Reply:
fixed